Store restored receipts in the storage

Longarithm opened the pull request:
**Idea**
~~Return `restored_receipts` from `apply_transactions` to indicate that some receipts were added in runtime due to migration, so they will not be lost in chain and could be accessed by Indexer.~~

Create new database migration to add receipts missing in store directly into it.

**Testing**
~~Extend existing `protocol_feature_restore_receipts_after_fix_tests` by running `get_receipt` method.~~

- setup archival mainnet node
- run `./target/release/neard --home ~/.near init`, observe migration:
```
$ ./target/debug/neard --home ~/.near run
Jul 09 12:57:59.635  INFO near: Version: 1.19.1, Build: c8e49bb5, Latest Protocol: 45
Jul 09 12:57:59.760  INFO near: Opening store database at "/home/Aleksandr1/.near/data"
apply_store_migrations
Jul 09 12:57:59.781  INFO near: Migrate DB from version 22 to 23
```
- stop node, access store using separate script and ensure that receipts are saved in it
```
for receipt in receipts {
    eprintln!("{}", receipt.get_hash());
    chain_store.get_receipt(&receipt.get_hash()).unwrap().unwrap();
}
```

bowenwang1996 commented on runtime/runtime/src/lib.rs:
```diff
@@ -122,6 +122,8 @@ pub struct ApplyResult {
     pub stats: ApplyStats,
     pub proof: Option<PartialStorage>,
+    /// Receipts restored during runtime apply after migration.
+    pub receipts_to_restore: Vec<Receipt>,
```
Why do we actually need this field? I thought we have receipts to restore as a global variable

Longarithm replied:
We do, but we need `chain_id` to decide to load it or not, which is stored on `Runtime` side.
And also at some moment in the past we decided that `Chain` should not know about `chain_id`, which discourages adding method like `get_chain_id` to `RuntimeAdapter`. So I leave `receipts_to_restore` computation to `Runtime`.

bowenwang1996 replied:
This is not a great solution. If possible, we should avoid introducing some field to `ApplyResult` only to use it once.

> also at some moment in the past we decided that Chain should not know about chain_id

I think it is more like we didn't see a good reason to do that. It does not mean it should never be done. Alternatively we can update it in `process_state_update` by writing directly to the database, which is also an ugly solution.

bowenwang1996 commented on chain/chain/src/chain.rs:
```diff
@@ -3709,6 +3712,9 @@ impl<'a> ChainUpdate<'a> {
                 receipt_proof_response.1,
             );
         }
+        if !apply_result.receipts_to_restore.is_empty() {
+            self.chain_store_update.save_receipts(&apply_result.receipts_to_restore);
+        }
```
I don't think we need this here. If we state sync past that specific block, we do not have the block in the first place and therefore it does not make sense for us to restore receipts here.

bowenwang1996 reviewed (commented):
Another option is that we can do a database migration to add those receipts

bowenwang1996 commented on nearcore/src/migrations.rs:
```diff
@@ -349,6 +349,20 @@ pub fn migrate_22_to_23(path: &String, near_config: &NearConfig) {
+/// Put receipts restored in scope of issue https://github.com/near/nearcore/pull/4248 to storage.
+pub fn migrate_23_to_24(path: &String, near_config: &NearConfig) {
+    let store = create_store(path);
+    if &near_config.genesis.config.chain_id == "mainnet" {
+        let genesis_height = near_config.genesis.config.genesis_height;
+        let mut chain_store = ChainStore::new(store.clone(), genesis_height);
+        let restored_receipts = serde_json::from_slice(&MAINNET_RESTORED_RECEIPTS)
+            .expect("File with receipts restored after apply_chunks fix have to be correct").get(&0u64);
+        let mut chain_store_update = ChainStoreUpdate::new(&mut chain_store);
+        chain_store_update.save_receipts(restored_receipts);
```
it is not committed

Longarithm commented:
Changed implementation to DB migration.

Note: added receipts will be stored in the store forever for each mainnet node, because they don't belong to any chunk and their refcount always equals 1. This is not ideal, but as long as file with these receipts takes only 300 KB of memory, it doesn't look like a big issue.

bowenwang1996 commented on nearcore/src/migrations.rs:
```diff
@@ -358,7 +360,12 @@ pub fn migrate_23_to_24(path: &String, near_config: &NearConfig) {
+            let bytes = receipt.try_to_vec().expect("Borsh cannot fail");
+            store_update.update_refcount(ColReceipts, receipt.get_hash().as_ref(), &bytes, 1);
+        }
+        store_update.commit().expect("");
```
```suggestion
        store_update.commit().unwrap();
```

bowenwang1996 replied:
Also committing per value is a bit excessive. You can commit at the end since we know that there are not that many receipts ti begin with.

Longarithm replied:
But I don't commit per value, commit is outside the loop.

bowenwang1996 commented on nearcore/src/migrations.rs:
```diff
@@ -349,6 +349,32 @@ pub fn migrate_22_to_23(path: &String, near_config: &NearConfig) {
+pub fn migrate_23_to_24(path: &String, near_config: &NearConfig) {
+    let store = create_store(path);
+    if &near_config.genesis.config.chain_id == "mainnet" {
+        let genesis_height = near_config.genesis.config.genesis_height;
+        let mut chain_store = ChainStore::new(store.clone(), genesis_height);
+        let restored_receipts: ReceiptResult = serde_json::from_slice(&MAINNET_RESTORED_RECEIPTS)
+            .expect("File with receipts restored after apply_chunks fix have to be correct");
+        let chain_store_update = ChainStoreUpdate::new(&mut chain_store);
+        let mut store_update = chain_store_update.store().store_update();
```
Why not construct a `StoreUpdate` directly from `store`?

Longarithm commented:
BTW it's a bit weird that `cargo fmt` changed files from runtime/runtime-params-estimator/src, because we have a test for it. 
cc @matklad?

matklad commented:
filed https://github.com/near/nearcore/issues/4505 for me to fix the formatter.

frol commented:
@bowenwang1996 we should have not stabilize `protocol_feature_restore_receipts_after_fix` before fixing #4337! Backport this PR to 1.20.x since otherwise we will break all indexers on mainnet!

cc @khorolets
